### ISSUE TITLE
MIR borrowck: wrong error for move into closure

### ISSUE THREAD

**Issue by vramana** [rust-lang/rust#46599]
```
# ramana at pc in ~/Documents/rust-lang/rust on git:fix-e0504 ✖︎ [18:05:33]
→ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc src/test/compile-fail/E0504.rs -Z borrowck=mir 
error[E0505]: cannot move out of `fancy_num` because it is borrowed
  --> src/test/compile-fail/E0504.rs:23:13
   |
21 |       let fancy_ref = &fancy_num;
   |                       ---------- borrow of `fancy_num` occurs here
22 | 
23 |       let x = move || {
   |  _____________^
24 | |         println!("child function: {}", fancy_num.num); //~ ERROR E0504
25 | |     };
   | |_____^ move out of `fancy_num` occurs here

error: aborting due to previous error


# ramana at pc in ~/Documents/rust-lang/rust on git:fix-e0504 ✖︎ [18:12:36]
→ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc src/test/compile-fail/E0504.rs                 
error[E0504]: cannot move `fancy_num` into closure because it is borrowed
  --> src/test/compile-fail/E0504.rs:24:40
   |
21 |     let fancy_ref = &fancy_num;
   |                      --------- borrow of `fancy_num` occurs here
...
24 |         println!("child function: {}", fancy_num.num); //~ ERROR E0504
   |                                        ^^^^^^^^^ move into closure occurs here

error: aborting due to previous error
```

cc @arielb1 

I am working on this.

**Comment by vramana** [comment]
Hi, 

I was little busy work stuff. I am taking a stab at this issue again today. I am using simpler case that generates less MIR code

```rs
struct FancyNum {
    num: u8,
}

fn main() {
    let mut fancy_num = FancyNum { num: 5 };
    // let fancy_ref = &fancy_num;

    let mut x = move || {
        fancy_num.num += 1;
    };

    x();
}
```


Generated MIR [here](https://play.rust-lang.org/?gist=30e2a8cccda0915629c95ca391bc2140&version=nightly)

When I uncomment the line `let fancy_ref = &fancy_num;`, this borrow conflits with this `_3 = move _1` in MIR above

```diff
       StorageLive(_2);                 // bb0[2]: scope 3 at src/main.rs:9:9: 9:14
       StorageLive(_3);                 // bb0[3]: scope 3 at src/main.rs:9:17: 11:6
       _3 = move _1;                    // bb0[4]: scope 3 at src/main.rs:9:17: 11:6
+      The line that causes the error.
       _2 = [closure@src/main.rs:9:17: 11:6] { fancy_num: move _3 }; // bb0[5]: scope 3 at src/main.rs:9:17: 11:6
                                         // closure
                                         // └ def_id: DefId(0/1:10 ~ playground[72d2]::main[0]::{{closure}}[0])
```


**Note**: _3 will be _4 in the debug log because of the extra variable fancy_ref (_2)

So the error occurs way before we enter the closure. 

https://github.com/rust-lang/rust/blob/16992930835ce3376a4aaed42307726e1fc78e45/src/librustc_mir/borrow_check/mod.rs#L349-L354

https://github.com/rust-lang/rust/blob/16992930835ce3376a4aaed42307726e1fc78e45/src/librustc_mir/borrow_check/mod.rs#L1016-L1023

https://github.com/rust-lang/rust/blob/16992930835ce3376a4aaed42307726e1fc78e45/src/librustc_mir/borrow_check/mod.rs#L850-L851


This call generates the error. The information we have is the place (_1) we access and the span (`scope 3 at src/main.rs:9:17: 11:6`) where we access it. How do I find the usages of _1 in this span?

**Comment by arielb1** [comment]
@vramana 

You need to find uses of `_1` within the closure. For that, you can follow down the MIR basic block until you find the closure statement, look at the closure def-id, and look for uses of the appropriate upvar within the closure body.

**Comment by vramana** [comment]
> For that, you can follow down the MIR basic block until you find the closure statement

I don't know how to implement this in code probably because I am not too familiar with API.

**Comment by vramana** [comment]
@arielb1 There is no `visit_lvalue` function. Can you explain your hints in a little more detail?

**Comment by arielb1** [comment]
@vramana 

Sorry for the delay. `lvalue` is now `place`, so `visit_lvalue` is now `visit_place`.

**Comment by nikomatsakis** [comment]
I wasn't able to observe this error, at least not with @vramana's minimized example:

https://play.rust-lang.org/?gist=6af331fc0c3dd1f455a2ecda1308991c&version=nightly

(I also tried removing the comment)

**Comment by nikomatsakis** [comment]
I guess this example gets an error:

```rust
#![feature(nll)]

struct FancyNum {
    num: u8,
}

fn main() {
    let mut fancy_num = FancyNum { num: 5 };
    let fancy_ref = &fancy_num;

    let mut x = move || {
        fancy_num.num += 1;
    };

    x();
    
    drop(fancy_ref);
}
```

**Comment by nikomatsakis** [comment]
I'm having trouble deciding how to categorize this -- I guess it's ultimately a "minor missing suggestion" and so belongs in the Release Candidate camp. Still, let me leave a few mentoring notes here. I'll leave a few notes on strategy, at least:

First, in the MIR borrowck, we have a move of some value and we need to see whether that value is being moved **into a closure**. This would be right around here, in the code:

https://github.com/rust-lang/rust/blob/860d169474acabdc53b9a698f8ce02eba7e0daeb/src/librustc_mir/borrow_check/error_reporting.rs#L172

We need to examine the MIR statement that performs the move (the location of which is identified by `context.loc`). If it is an assignment where the right hand side is an [`Rvalue::Aggregate`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/enum.Rvalue.html#variant.Aggregate), and the `def_id` of the aggregate is a closure (you can test with [`tcx.is_closure(def_id)`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.is_closure)), then indeed the move is occuring as part of closure construction.

In that case, you can find the index of the move amongst the arguments of the aggregate. That will be the upvar index. We can check the result of [the `freevars(def_id)` query](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.freevars) for the closure's def-id. Each [`FreeVar`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/struct.Freevar.html) describes one variable captured by the closure; the vector has the same ordering as the aggregate fields. The field `span` of this struct is the span where the value is used within the closure:

https://github.com/rust-lang/rust/blob/860d169474acabdc53b9a698f8ce02eba7e0daeb/src/librustc/hir/mod.rs#L2220-L2221

So we should extract that value and add the label on *that* span instead.
